### PR TITLE
AY-7217_Add support to sync task assignees and statuses for all entities

### PR DESCRIPTION
**Pull request by fabiaserra**
## Changelog Description
This adds support to sync task assignees to AYON tasks and statuses on all entities (including Versions)

## Testing notes:
1. synchronize users first
chrome_wk2mdKuBG1
2. Assign users to at least 1 access group in `manageProjects/userSettings?project=YOUR_PROJECT`
    without it sync of assignees from SG>AYON wont work, but will not fail. Assigning users to AYON project must be implemented during event based user synch.
2. experiment with adding/removing assignee/tags/status in AYON (right column in `Editor`)
chrome_gQcwKQRGad

3. do the same on SG side

(^edited by kalisp)

## original PR description kept for posterity
This adds support to sync task assignees to AYON tasks and statuses on all entities (including Versions)

This requires both of these PRs:
* https://github.com/ynput/ayon-python-api/pull/157
* https://github.com/ynput/ayon-python-api/pull/158

AY-7217

### CODE REVIEW

File: services/shotgrid_common/ayon_shotgrid_hub/update_from_ayon.py [hunk 1]
```diff
@@ -386,6 +386,7 @@ def _create_sg_entity(
             or ay_entity.entity_type != "task"
             and ay_entity.folder_type == "AssetCategory"
     ):
+        # TODO: why not??
```
[the remark this answers is not in the document]
Comment (jakubjezek001) [reply services/shotgrid_common/ayon_shotgrid_hub/update_from_ayon.py:389]: Our SG entity currently lacks an Asset category; it's inferred solely from the sg_asset_type of SG Assets. While we facilitate Asset category synchronization through this conversion, the AssetCategory folder type is exclusive to AYON and should be omitted here.

Could your server be treating it as a custom entity, or can you direct me to documentation that specifies the entity type?

File: services/shotgrid_common/ayon_shotgrid_hub/update_from_shotgrid.py [hunk 2]
```diff
@@ -342,9 +404,10 @@ def remove_ayon_entity_from_sg_event(
     """
     # for now we are ignoring Task type entities
     # TODO: Handle Task entities
-    if sg_event["entity_type"] == "Task":
-        log.info("Ignoring Task entity.")
-        return
+    # Why?
+    # if sg_event["entity_type"] == "Task":
+    #     log.info("Ignoring Task entity.")
+    #     return
```
Comment (jakubjezek001) [inline services/shotgrid_common/ayon_shotgrid_hub/update_from_shotgrid.py:410]: Indeed, clearer communication was needed. A glitch in the AYON server caused the blockage. @fabiaserra, did it work when you tested it? Which server version are you using?

File: services/shotgrid_common/ayon_shotgrid_hub/update_from_ayon.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabiaserra) [reply services/shotgrid_common/ayon_shotgrid_hub/update_from_ayon.py:389]: You are right yeah, I realized that was true after and I forgot I never updated this conversation

File: services/shotgrid_common/utils.py [hunk 3]
```diff
@@ -672,6 +679,24 @@ def get_sg_entities(
...
+                task_assignees_list = []
+                for assignee in task_assignees:
+                    # Skip task assignments that aren't from a human user (i.e. groups)
+                    # TODO: add support for group assignments
+                    if assignee["type"] != "HumanUser":
+                        continue
+                    sg_user = get_sg_user_by_id(
+                        sg_session, assignee["id"], extra_fields=["login"]
+                    )
+                    task_assignees_list.append(sg_user["login"])
```
Comment (jakubjezek001) [inline services/shotgrid_common/utils.py:696]: Here we will need to refactor the code a little so we are actually getting AY user names based on `sg_user_id` key store in AY user entity.data. Look here #76 .

File: services/shotgrid_common/utils.py [hunk 4]
```diff
@@ -734,6 +759,24 @@ def get_sg_entity_as_ay_dict(
...
+        task_assignees_list = []
+        for assignee in task_assignees:
+            # Skip task assignments that aren't from a human user (i.e. groups)
+            # TODO: add support for group assignments
+            if assignee["type"] != "HumanUser":
+                continue
+            sg_user = get_sg_user_by_id(
+                sg_session, assignee["id"], extra_fields=["login"]
+            )
+            task_assignees_list.append(sg_user["login"])
```
Comment (jakubjezek001) [inline services/shotgrid_common/utils.py:777]: The same as https://github.com/ynput/ayon-shotgrid/pull/113/files#r1863570202

**Comment by kalisp** [comment]
Fixed conflicts, translated SG user ids to AYON users and back.

Tested synchronizations from SG>AYON and back, it seems to work.

File: server/__init__.py [hunk 5]
```diff
@@ -110,3 +121,29 @@ async def create_shotgrid_attributes(self) -> bool:
         else:
             logging.debug("Shotgrid Attributes already exist.")
             return False
+
+    async def get_user_by_sg_id(
```
Comment (iLLiCiTiT) [inline server/__init__.py:125]: Shoulnd't this return just username?

File: services/shotgrid_common/ayon_shotgrid_hub/__init__.py [hunk 6]
```diff
@@ -434,7 +436,13 @@ def react_to_ayon_event(self, ayon_event):
                     self._ay_project,
                     self.custom_attribs_map,
                 )
-            case "entity.task.status_changed" | "entity.folder.status_changed" | "entity.task.tags_changed" | "entity.folder.tags_changed":
+            case (
+                "entity.task.status_changed" |
+                "entity.folder.status_changed" |
+                "entity.task.tags_changed" |
+                "entity.folder.tags_changed" |
+                "entity.task.assignees_changed"
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/ayon_shotgrid_hub/__init__.py:444]:
```suggestion
                "entity.task.status_changed"
                | "entity.folder.status_changed"
                | "entity.task.tags_changed"
                | "entity.folder.tags_changed"
                | "entity.task.assignees_changed"
```

File: services/shotgrid_common/utils.py [hunk 7]
```diff
@@ -117,6 +117,7 @@ def _sg_to_ay_dict(
     ay_entity_type = "folder"
     task_type = None
     folder_type = None
+    exception_attribs = {"status", "assignees", "tags"}
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:120]: I would use something like `root_attributes`. The `exception_attribs` sounds "different" from what it is.

File: services/shotgrid_common/utils.py [hunk 8]
```diff
@@ -1359,6 +1411,41 @@ def create_new_ayon_entity(
...
+                ay_entity.status = new_status_name
+            except ValueError as e:
+                # `ValueError: Status ip is not available on project.`
+                # NOTE: this doesn't really raise exception?
+                log.warning(f"Status sync not implemented: {e}")
+
+    assignees = sg_ay_dict.get("assignees")
+    if assignees:
+        try:
+            # INFO: it was causing error so trying to set status directly
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1440]: Wrong comment, and probably unnecessary try except block (copy paste?)

File: services/shotgrid_common/utils.py [hunk 9]
```diff
@@ -1359,6 +1411,41 @@ def create_new_ayon_entity(
...
     )
 
+    status = sg_ay_dict.get("status")
+    if status:
+        # Entity hub expects the statuses to be provided with the `name` and
+        # not the `short_name` (which is what we get from SG) so we convert
+        # the short name back to the long name before setting it
+        status_mapping = {
+            status.short_name: status.name for status in
+            entity_hub.project_entity.statuses
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1421]:
```suggestion
            status.short_name: status.name
            for status in entity_hub.project_entity.statuses
```

File: services/shotgrid_common/utils.py [hunk 10]
```diff
@@ -1359,6 +1411,41 @@ def create_new_ayon_entity(
...
+        }
+        new_status_name = status_mapping.get(status)
+        if not new_status_name:
+            log.warning(
+                "Status with short name '%s' doesn't exist in project", status
+            )
+        else:
+            try:
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1429]: You can check for scope of the status before you set it.

File: services/shotgrid_common/utils.py [hunk 11]
```diff
@@ -1359,6 +1411,41 @@ def create_new_ayon_entity(
...
+                # INFO: it was causing error so trying to set status directly
+                ay_entity.status = new_status_name
+            except ValueError as e:
+                # `ValueError: Status ip is not available on project.`
+                # NOTE: this doesn't really raise exception?
+                log.warning(f"Status sync not implemented: {e}")
+
+    assignees = sg_ay_dict.get("assignees")
+    if assignees:
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1438]: Removement of assignees should be propagated too, or?
```suggestion
    if assignees is None:
```

Comment (kalisp) [reply services/shotgrid_common/utils.py:1438]: This is creation of new User in AYON, there is no resetting here.

**Review by jakubjezek001 — APPROVED** [review]
LGTM

File: services/shotgrid_common/utils.py [hunk 12]
```diff
@@ -1359,6 +1411,37 @@ def create_new_ayon_entity(
...
+                "Status with short name '%s' doesn't exist in project", status
+            )
+        else:
+            try:
+                # INFO: it was causing error so trying to set status directly
+                ay_entity.status = new_status_name
+            except ValueError as e:
+                # `ValueError: Status ip is not available on project.`
+                # NOTE: this doesn't really raise exception?
+                log.warning(f"Status sync not implemented: {e}")
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1435]: But there is missing entity type...
```suggestion
        status_mapping = {
            status.short_name: status
            for status in entity_hub.project_entity.statuses
        }
        new_status = status_mapping.get(status)
        if new_status is None:
            log.warning(
                "Status with short name '%s' doesn't exist in project", status
            )
        elif entity_type not in new_status.scope:
            ay_entity.status = new_status.name
        else:
            log.warning(f"Status {status} is not available on {entity_type}")
```

File: services/shotgrid_common/utils.py (hunk 11 above)
Comment (iLLiCiTiT) [reply services/shotgrid_common/utils.py:1438]: Not creation of new "User", but I understand... Make sense.

File: services/shotgrid_common/utils.py [hunk 13]
```diff
@@ -1373,3 +1456,54 @@ def create_new_ayon_entity(
...
+
+    task_assignees_list = []
+    for assignee in task_assignees:
+        # Skip task assignments that aren't from a human user (i.e. groups)
+        # TODO: add support for group assignments
+        if assignee["type"] != "HumanUser":
+            continue
+        ayon_user_name = get_ayon_name_by_sg_id(assignee["id"])
+        if not ayon_user_name:
+            log.warning(f"Didn't find user for '{assignee['id']}'")
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1506]:
```suggestion
            log.warning(f"Didn't find user for '{assignee['id']}'")
            continue
```

File: services/shotgrid_common/utils.py [hunk 14]
```diff
@@ -1277,27 +1320,36 @@ def update_ay_entity_custom_attributes(
...
             ay_entity.tags = [tag["name"] for tag in attrib_value]
         elif ay_attrib == "status":
-            # TODO: Implement status update
+            # Entity hub expects the statuses to be provided with the `name` and
+            # not the `short_name` (which is what we get from SG) so we convert
+            # the short name back to the long name before setting it
+            status_mapping = {
+                status.short_name: status.name for status in ay_project.statuses
+            }
+            new_status_name = status_mapping.get(attrib_value)
```
Comment (iLLiCiTiT) [inline services/shotgrid_common/utils.py:1343]:
```suggestion
            status_mapping = {
                status.short_name: status
                for status in ay_project.statuses
            }
            new_status = status_mapping.get(attrib_value)
            if ay_entity.entity_type in new_status.scope:
                ay_entity.status = new_status.name
            else:
                logging.warning(
                    f"Status '{attrb_value}' not available"
                    f" for {entity.entity_type}"
                )
```

**Comment by kalisp** [comment]
Let test more before release.

Thanks Fabia for contribution, even if it took a for while to review.

**Pull request merged**